library/base.go

```diff
@@ -0,0 +1,227 @@
+package library
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+type runningProcess struct {
+	cmd     *exec.Cmd
+	sender  Sender
+	output  *SafeBuffer
+	tracker *StatusTracker
+	done    <-chan bool
+	pid     int
+	tempdir string
+}
+
+type RunnerImpl struct {
+	running     map[Process]*runningProcess
+	authz       AuthZRules
+	wrapperPath string
+	mtx         sync.Mutex
+	cgi         CgroupImpl
+}
+
+func NewRunner(a AuthZRules, cgi CgroupImpl) *RunnerImpl {
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	wrapper := fmt.Sprintf("%s/../wrapper/wrapper", cwd)
+	wrapper = path.Clean(wrapper)
+
+	return &RunnerImpl{
+		running:     map[Process]*runningProcess{},
+		wrapperPath: wrapper,
+		cgi:         cgi,
+		authz:       a,
+	}
+}
+
+var _ Runner = (*RunnerImpl)(nil)
+
+func (a *AuthZRules) getControlGroup(name string) (*ControlGroup, error) {
+	for _, cg := range a.ControlGroups {
+		if cg.Name == name {
+			return &cg, nil
+		}
+	}
+	return nil, errors.New("no such control group name")
```

Comment:
Consider adding the name to the error message to make it easier to understand what went wrong?
```suggestion
	return nil, fmt.Errorf("control group %v not found", name)
```